modules/application-server-commons/src/main/java/org/testcontainers/applicationserver/ApplicationServerContainer.java

```diff
@@ -0,0 +1,415 @@
+     * for a successful response from the {@link ApplicationServerContainer#getApplicationURL()}.
+     *
+     * This method will overwrite the path used by the HttpWaitStrategy but will keep the httpPort.
+     *
+     * Calling this method more than once will replace the current readinessPath if it was already set.
+     *
+     * @param readinessPath - The path to be polled for readiness.
+     * @return self
+     */
+    public ApplicationServerContainer withReadiness(String readinessPath) {
+        return withReadiness(readinessPath, this.httpPort);
+    }
+
+    /**
+     * By default, the ApplicationServerContainer will be configured with a HttpWaitStrategy and wait
+     * for a successful response from the {@link ApplicationServerContainer#getApplicationURL()}.
+     *
+     * This method will overwrite the path used by the HttpWaitStrategy.
+     * This method will overwrite the port used by the HttpWaitStrategy.
+     *
+     * Calling this method more than once will replace the current readinessPath and readinessPort if any were already set.
+     *
+     * @param readinessPath - The path to be polled for readiness.
+     * @param readinessPort - The port that should be used for the readiness check.
+     * @return self
+     */
+    public ApplicationServerContainer withReadiness(@NonNull String readinessPath, @NonNull Integer readinessPort) {
+        this.readinessPath = normalizePath(readinessPath);
+        this.readinessPort = readinessPort;
+        return this;
+    }
+
+    /**
+     * Set the timeout configured on the HttpWaitStrategy.
+     *
+     * Calling this method more than once will replace the current httpWaitTimeout if it was already set.
+     * Default value is 60 seconds
+     *
+     * @param httpWaitTimeout - The duration of time to wait for a successful http response
+     * @return self
+     */
+    public ApplicationServerContainer withHttpWaitTimeout(Duration httpWaitTimeout) {
+        this.httpWaitTimeout = httpWaitTimeout;
+        return this;
+    }
+
+    /**
+     * An environment variable whose value needs to wait until the configuration stage to be evaluated.
+     * The common use case for this would be to pass inter-container connection data.
+     *
+     * For example an application container often needs to have connection data to a dependent database:
+     * <pre>
+     *     OracleContainer db = new OracleContainer(...).withExposedPorts(1521);
+     *     ApplicationContainer app = new ApplicationContainer(...)
+     *       .withLazyEnv( "oracle.url", () -&gt; db.getJdbcUrl() ) // Need to wait until 'db' container is configured
+     *       .dependsOn(db);
+     * </pre>
+	 *
+     * @param key   - the key
+     * @param valueSupplier - the function that supplies the value
+     * @return self
+     */
+    public ApplicationServerContainer withLazyEnv(String key, Supplier<String> valueSupplier) {
+        lazyEnvVars.put(key, valueSupplier);
+        return this;
+    }
```

Comment:
The problem I have with this is that I have to have internal knowledge of how the PostgreSQLContainer configures itself. I would like to be able to do this: 

```diff
-        .withEnv("DB_USER", "test")
-        .withEnv("DB_PASSWORD", "test")
-        .withEnv("DB_JDBC_URL", "jdbc:postgresql://postgres:5432/test")

+        .withEnv("DB_USER", postgres.getUsername())
+        .withEnv("DB_PASSWORD", postgres.getPassword())
+        .withEnv("DB_JDBC_URL", postgres.getJdbcUrl())
```

Which doesn't work since the getJdbcUrl will still call getMappedPort

Reply:
`PostgreSQLContainer` provides useful methods to override the username, password and databaseName. So, that way it should not be that hard to know the setup.

I understand what you mentioned but I think it's a different topic to discuss. Let's avoid of solving it adding `withLazyEnv` in `ApplicationServerContainer` when the "issue" is broader. WDYT?

Reply:
Okay, I have removed this and will open a seperate issue to address this in a more general way.